src/Hcal/HcalDigiProducer.cxx

```diff
@@ -1,392 +1,235 @@
-      // ldmx::HcalID::LEFT)   cur_ypos =  half_total_width/2.0 -
-      // ecal_width/4.0; if (cur_subsection == ldmx::HcalID::RIGHT)  cur_ypos =
-      // -half_total_width/2.0 + ecal_width/4.0;
-
-      // This would be the quantized z position. The side_hcal_z0 value must be
-      // derived fromn the geometry! float side_hcal_z0(215.5); cur_zpos =
-      // side_hcal_z0+(cur_strip+0.5)*strip_width ;
-
-      // This is the quantized position along the thickness of the bar - check
-      // RIGHT / LEFT float back_hcal_layer_thickness(39); float
-      // side_hcal_xy_offset(294); if (cur_subsection == ldmx::HcalID::TOP)
-      // cur_ypos = side_hcal_xy_offset+(cur_layer-1)*back_hcal_layer_thickness;
-      // if (cur_subsection == ldmx::HcalID::BOTTOM) cur_ypos =
-      // -side_hcal_xy_offset-(cur_layer-1)*back_hcal_layer_thickness; if
-      // (cur_subsection == ldmx::HcalID::LEFT)   cur_xpos =
-      // -side_hcal_xy_offset-(cur_layer-1)*back_hcal_layer_thickness; if
-      // (cur_subsection == ldmx::HcalID::RIGHT)  cur_xpos =
-      // side_hcal_xy_offset+(cur_layer-1)*back_hcal_layer_thickness;
     }
+  }
 
-    if (hcalLayerPEs[detIDraw] >= readoutThreshold_) {
-      ldmx::HcalHit hit;
-      hit.setID(detIDraw);
-      hit.setPE(hcalLayerPEs[detIDraw]);
-      hit.setMinPE(hcalLayerMinPEs[detIDraw]);
-      hit.setAmplitude(hcalLayerPEs[detIDraw]);
-      hit.setEnergy(energy);
-      hit.setTime(hcaldetIDTime[detIDraw]);
-      hit.setXPos(cur_xpos);  // quantized and smeared positions
-      hit.setYPos(cur_ypos);  // quantized and smeared positions
-      hit.setZPos(cur_zpos);
-      hit.setNoise(false);
-
-      hcalRecHits.push_back(hit);
+  /******************************************************************************************
+   * Noise Simulation on Empty Channels
+   *****************************************************************************************/
+  if (noise_) {
+    int numChannels = 0;
+    for (int l = 0; l < hcalGeometry.getNumSections(); l++) {
+      int nChannels =
+          hcalGeometry.getNumLayers(l) * hcalGeometry.getNumStrips(l);
+      if (l == 0) nChannels *= 2;
+      numChannels += nChannels;
     }
+    int numEmptyChannels = numChannels - hcalDigis.getNumDigis();
+    // noise generator gives us a list of noise amplitudes [mV] that randomly
+    // populate the empty channels and are above the readout threshold
+    auto noiseHitAmplitudes{
+        noiseGenerator_->generateNoiseHits(numEmptyChannels)};
+    std::vector<double> voltages(1, 0.), times(1, 0.);
+    for (double noiseHit : noiseHitAmplitudes) {
+      // generate detector ID for noise hit
+      // making sure that it is in an empty channel
+      unsigned int noiseID;
+      do {
+        int sectionID = noiseInjector_->Integer(hcalGeometry.getNumSections());
+        int layerID = noiseInjector_->Integer(hcalGeometry.getNumLayers(0));
+        int stripID = noiseInjector_->Integer(hcalGeometry.getNumStrips(0));
```

Comment:
Doesn't the number of layers and strips depend on the section? You just determined a section number, so use it to make sure the generated ID is a valid ID.

Reply:
Oops you are right, modified [here](https://github.com/LDMX-Software/Hcal/blob/1857ada0d4642626ff99759400ce6525f049a663/src/Hcal/HcalDigiProducer.cxx#L267-L270).